### DISCUSSION TITLE
Does scikit-survival models supported by SHAP or any other explainable libraries ?

### DISCUSSION THREAD

**Discussion by sagardatascientists** [sebp/scikit-survival#213] [Ideas]
Hello All, I am investigating to call explainable models like SHAP, LIME into Scikit-survival models and wondering has anyone tried it before or not ?
It would be great to know your ideas and thoughts around it.

Thanks,
Sagar

**Comment by sebp** [comment]
In principle, those methods could be used to explain survival models as well. In the case of SHAP, I know that they have generic explainers that treat the model as a black-box, and specific explainers that leverage the internals of learners, such as tree-based learners. The former should work out-of-the-box, the latter might rely on certain properties that don't apply to Survival trees. Note that this isn't a definite answer, just some of my thoughts without looking into the details of the algorithms.

**Reply by psteinb** [reply]
Would be cool to have an example somewhere. Let me give the generic SHAP explainer a try. One thing, I can already say, a sksurv model within a sklearn pipeline doesn't work. :weary:

**Reply by psteinb** [reply]
I just checked using the random-surival-forest notebook contained in the `doc/user_guide/` folder.
Neither:
```
expl = shap.Explainer(rsf)
```
nor 
```
expl = shap.TreeExplainer(rsf)
```
works. I used
```
Name: shap
Version: 0.39.0
Name: scikit-survival
Version: 0.15.0.post0
```

**Comment by sebp** [comment]
What do you mean by "does not work" mean? Do you get an exception or the 
SHAP values aren't what you expected?

**Reply by psteinb** [reply]
I am super sorry. 

```
---------------------------------------------------------------------------
Exception                                 Traceback (most recent call last)
/tmp/ipykernel_52797/2067042318.py in <module>
----> 1 expl = shap.Explainer(rsf)

~/foo/venv/bar/lib64/python3.9/site-packages/shap/explainers/_explainer.py in __init__(self, model, masker, link, algorithm, output_names, feature_names, **kwargs)
    166                 # if we get here then we don't know how to handle what was given to us
    167                 else:
--> 168                     raise Exception("The passed model is not callable and cannot be analyzed directly with the given masker! Model: " + str(model))
    169 
    170             # build the right subclass

Exception: The passed model is not callable and cannot be analyzed directly with the given masker! Model: RandomSurvivalForest(max_features='sqrt', min_samples_leaf=15,
                     min_samples_split=10, n_estimators=1000, n_jobs=-1,
                     random_state=20)
```

and 

```
---------------------------------------------------------------------------
Exception                                 Traceback (most recent call last)
/tmp/ipykernel_52797/1379721347.py in <module>
----> 1 expl = shap.TreeExplainer(rsf)

~/projects/haicu/hecktor/2021/repo/rsc/venv/hecktor/lib64/python3.9/site-packages/shap/explainers/_tree.py in __init__(self, model, data, model_output, feature_perturbation, feature_names, **deprecated_options)
    145         self.feature_perturbation = feature_perturbation
    146         self.expected_value = None
--> 147         self.model = TreeEnsemble(model, self.data, self.data_missing, model_output)
    148         self.model_output = model_output
    149         #self.model_output = self.model.model_output # this allows the TreeEnsemble to translate model outputs types by how it loads the model

~/foo/venv/bar/lib64/python3.9/site-packages/shap/explainers/_tree.py in __init__(self, model, data, data_missing, model_output)
    975             self.base_offset = model.init_params[param_idx]
    976         else:
--> 977             raise Exception("Model type not yet supported by TreeExplainer: " + str(type(model)))
    978 
    979         # build a dense numpy version of all the tree objects

Exception: Model type not yet supported by TreeExplainer: <class 'sksurv.ensemble.forest.RandomSurvivalForest'>
```

**Reply by psteinb** [reply]
For your interest, here is the notebook as adapted from the repo.
https://gist.github.com/psteinb/d04ed92eccbb2b34e9e50b5ba817a22b

**Comment by sebp** [comment]
My understanding is that you don't pass rsf directly, but a callable, 
which I assume should be rsf.predict

**Comment by psteinb** [comment]
Ok, based on the notebook, I played around a bit more. Apparently the following is able to run at the end of [random-survival-forest.ipynb](https://github.com/sebp/scikit-survival/blob/master/doc/user_guide/random-survival-forest.ipynb):

``` python
import shap

explainer = shap.Explainer(rsf.predict, X_train, feature_names=feature_names)
shaps = explainer(X_test)
```

@sebp thanks for reminding me about the `rsf.predict` trick.

**Reply by psteinb** [reply]
here is the result using a [waterfall](https://shap.readthedocs.io/en/latest/generated/shap.plots.waterfall.html#shap-plots-waterfall) plot:

``` python
shap.plots.waterfall(shaps[0])
```

![fp_plain](https://user-images.githubusercontent.com/1465603/128009109-3190499b-3ea1-41fb-b432-184acbdc07fa.jpg)

Interesting to see the different to the permutationimportance:

![image](https://user-images.githubusercontent.com/1465603/128009200-cd507fdc-de0c-445c-a7f2-571412fbec17.png)

**Reply by hyejinhahihong** [reply]
hello @sebp and @psteinb . your answer and question helped me a lot cos i was having a same problem. 
However, i wanted the shap values using rsf.predict_survival_function instead of rsf.predict method. 

So i tried the following code
```python
import shap

explainer = shap.Explainer(rsf.predict_survival_function, X_train, feature_names=feature_names)
shaps = explainer(X_test)
```

but it gives me and error that TypeError : float() argument must be a string or a number, not 'Stepfunction'

in this case, what can be the solution?

**Reply by sebp** [reply]
shap only works with scalar predictions, not functions. For Cox models, explaining the output of `predict`  implicitly explains the survival functions too. For other other estimators (e.g. RandomSurvivalForest) you would need to run `predict_survival_function`, pick a specific time point `t`, and evaluate the predicted survival functions (instances of `Stepfunction`) at that time point `t`. This yields the desired scalar value (probability of remaining event-free beyond time point `t`) that shap can deal with.

**Comment by sebp** [comment]
Well, permutation-based feature importances are global measures, whereas 
SHAP is a local, i.e. per patient, measure. You would need to average 
absolute SHAP values across all patients to get a fair comparison.

**Reply by psteinb** [reply]
Sorry to nag on this, a beeswarm view for a more global view. disclaimer: I am exploring the `shap` library here.

![beeswarm_plain](https://user-images.githubusercontent.com/1465603/128016749-9efe8504-f4d3-4ad5-b0cb-7814750d9f60.jpg)

**Reply by YounessAzimzade** [reply]
> Sorry to nag on this, a beeswarm view for a more global view. disclaimer: I am exploring the `shap` library here.
> 
> ![beeswarm_plain](https://user-images.githubusercontent.com/1465603/128016749-9efe8504-f4d3-4ad5-b0cb-7814750d9f60.jpg)

Hey, 
did you plot this for your rsf mode? if so, how did you extract shap values from the explainer? 
thanks

**Comment by DylanGuo1101** [comment]
Hi, can I use shap_interaction_values to calculate the shap interaction value of scikit-survival models?

**Reply by sebp** [reply]
I think so. If you encounter problems, feel free to report back.

**Reply by DylanGuo1101** [reply]
Thank you for your reply, I encountered a compatibility-related error when calculating SHAP interaction values.


```
# SHAP analysis for Ridge Cox, training

explainer = shap.Explainer(coxph.predict, x_train_norm.values, feature_names=features, seed=None)
shap_values = explainer(x_train_norm)
shap_values_test = explainer(x_test_norm)

# Successfully generated swarm Plot
shap.summary_plot(shap_values, x_train_norm, feature_names=features, show = False)

# Try to calculate the interaction value, but an error is reported
shap_interaction_values = shap_values_train.shap_interaction_values(coxph)

---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[6], line 1
----> 1 shap_values_train.shap_interaction_values(coxph)

AttributeError: 'Explanation' object has no attribute 'shap_interaction_values'
```

```
Environment Info:
scikit-survival version: 0.24.0
SHAP version: 0.45.1
```

**Comment by sebp** [comment]
I believe you need to call explainer.shap_interaction_values

See 
https://shap-community.readthedocs.io/en/latest/example_notebooks/tabular_examples/tree_based_models/Basic%20SHAP%20Interaction%20Value%20Example%20in%20XGBoost.html#SHAP-Interaction-Values

**Reply by DylanGuo1101** [reply]
In fact, I did as the official documents do. 
😭😭 
<img width="1164" height="804" alt="082b7379-7018-49f5-9c9c-f31bc57fe946" src="https://github.com/user-attachments/assets/d0799a3a-bfee-48b5-ae1b-981e4996029d" />
